adapter/src/ethereum.rs

```diff
@@ -1,101 +1,505 @@
+        let (decoded_adress, decoded_signature) = match (hex::decode(signer), hex::decode(sig)) {
+            (Ok(address), Ok(sig)) => (address, sig),
+            (_, _) => {
+                return Err(AdapterError::Signature(
+                    "invalid signature or address".to_string(),
+                ))
+            }
         };
 
-        Ok(is_same)
+        let address = Address::from_slice(&decoded_adress);
+        let signature = Signature::from_electrum(&decoded_signature);
+        let message = Message::from_slice(&hash_message(state_root));
+
+        match verify_address(&address, &signature, &message) {
+            Ok(result) => Ok(result),
+            Err(_) => Ok(false),
+        }
     }
 
-    fn validate_channel(&self, _channel: &Channel) -> AdapterResult<bool> {
-        // @TODO
+    fn validate_channel(&self, channel: &Channel) -> AdapterResult<bool> {
+        let eth_channel: EthereumChannel = channel.into();
+        let channel_id = eth_channel
+            .hash_hex(&self.config.ethereum_core_address)
+            .map_err(|_| map_error("Failed to hash the channel id"))?;
+
+        if channel_id != channel.id {
+            return Err(AdapterError::Configuration(
+                "channel.id is not valid".to_string(),
+            ));
+        }
+
+        if check_validator_id_checksum(channel) {
+            return Err(AdapterError::Configuration(
+                "channel.validators: all addresses are checksummed".to_string(),
+            ));
+        }
+        // check if channel is valid
+        let is_channel_valid = EthereumAdapter::is_channel_valid(&self.config, channel);
+        if is_channel_valid.is_err() {
+            return Err(AdapterError::InvalidChannel(
+                is_channel_valid
+                    .err()
+                    .expect("failed to get channel error")
+                    .to_string(),
+            ));
+        }
+        // query the blockchain for the channel status
+        let contract_address = Address::from_slice(self.config.ethereum_core_address.as_bytes());
+        let contract = get_contract(&self.config, contract_address, &ADEXCORE_ABI)
+            .map_err(|_| map_error("failed to init core contract"))?;
+
+        let channel_status: U256 = contract
+            .query("states", channel_id, None, Options::default(), None)
+            .wait()
+            .map_err(|_| map_error("contract channel status query failed"))?;
+
+        if channel_status != 1.into() {
+            return Err(AdapterError::Configuration(
+                "channel is not Active on the ethereum network".to_string(),
+            ));
+        }
+
         Ok(true)
     }
 
-    fn session_from_token(&self, _token: &str) -> AdapterResult<String> {
-        // @TODO
-        Ok("hello".to_string())
+    fn session_from_token(&self, token: &str) -> AdapterResult<Session> {
+        let token_id = token.to_owned()[token.len() - 16..].to_string();
```

Comment:
This can panic, if the token passed doesn't fit in the range: `[token.len() - 16..].` (i.e. token is less than 16 chars)
IMO this should panic it should return an `Err` in that case.

Reply:
yeah, fixed